### ISSUE TITLE
Question: Why vanilla convolution performs better than B2CNet modules?

### ISSUE THREAD

**Issue by mengxin3216** [bao11seven/B2CNet#4]
When reproducing B2CNet on the WHU-CD dataset using the Open-CD framework, I replaced your three designed modules with vanilla convolutions and simple difference operations. Surprisingly, the modified model achieves **higher F1 score (around 93.1%) than the original B2CNet (about 92.1%) with fewer parameters. I am confused why replacing the proposed modules with basic convolution leads to performance improvement rather than degradation. Could you please let me know if I missed any key implementation details or training settings in your original work? Any explanations or suggestions would be greatly appreciated.

from backbone import resnet18
import torch
import torch.nn as nn

def initialize_weights(*models):
    for model in models:
        for module in model.modules():
            if isinstance(module, nn.Conv2d) or isinstance(module, nn.Linear):
                nn.init.kaiming_normal_(module.weight)
                if module.bias is not None:
                    module.bias.data.zero_()
            elif isinstance(module, nn.BatchNorm2d):
                module.weight.data.fill_(1)
                module.bias.data.zero_()

class FC(nn.Module):
    """
    Feature Concatenation
    用简单拼接替代 SDIM / DDIM
    输入输出接口与 DDIM 对齐
    """
    def __init__(self, inp, out, size=None):
        super(FC, self).__init__()

        mid_c = inp // 2   # 与单时相通道一致

        self.reduce = nn.Sequential(
            nn.Conv2d(inp, mid_c, kernel_size=1, bias=False),
            nn.BatchNorm2d(mid_c),
            nn.ReLU(inplace=True)
        )

        self.out = nn.Sequential(
            nn.Conv2d(mid_c, out, kernel_size=3, padding=1, bias=False),
            nn.BatchNorm2d(out),
            nn.ReLU(inplace=True)
        )

    def forward(self, inp1, inp2, last_feature=None):
        x = torch.cat([inp1, inp2], dim=1)   # [B, inp, H, W]
        feat = self.reduce(x)                # [B, inp//2, H, W]

        if last_feature is not None:
            feat = feat + last_feature

        out = self.out(feat)
        return out

class SimpleFusion(nn.Module):
    """
    baseline中的简单融合
    拼接 + 1x1卷积 + 上采样，用于替代 defm
    """
    def __init__(self, inc, outc):
        super(SimpleFusion, self).__init__()
        self.conv = nn.Conv2d(inc, outc, kernel_size=1, bias=False)
        self.Up = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)

    def forward(self, diff, accom):
        out = torch.cat([accom, diff], dim=1)
        out = self.conv(out)
        out = self.Up(out)
        return out

class FD(nn.Module):
    """
    Feature Difference
    用最直接的绝对差分替代 cbm
    输入输出尺寸与 cbm 保持一致
    """
    def __init__(self, in_channel):
        super(FD, self).__init__()
        self.proj = nn.Sequential(
            nn.Conv2d(in_channel, in_channel, kernel_size=1, bias=False),
            nn.BatchNorm2d(in_channel),
            nn.ReLU(inplace=True)
        )

    def forward(self, x1, x2):
        diff = torch.abs(x1 - x2)
        out = self.proj(diff)
        return out

class B2CNet(nn.Module):
    def __init__(self, pretrained=True):
        super().__init__()
        self.reset = resnet18(pretrained=pretrained)

        self.bfam_1 = FC(128, 32)
        self.bfam_2 = FC(256, 64)
        self.bfam_3 = FC(512, 128, (32, 32))
        self.bfam_4 = FC(1024, 256, (16, 16))

        self.cbm_1 = FD(64)
        self.cbm_2 = FD(128)
        self.cbm_3 = FD(256)
        self.cbm_4 = FD(512)

        self.defm_1 = SimpleFusion(64, 32)
        self.defm_2 = SimpleFusion(128, 64)
        self.defm_3 = SimpleFusion(256, 128)
        self.defm_4 = SimpleFusion(512, 256)

        self.D_Conv_4 = nn.Sequential(nn.Conv2d(512, 256, kernel_size=1),
                                      nn.BatchNorm2d(256),
                                      nn.ReLU(inplace=True)
                                      )
        self.D_Conv_3 = nn.Sequential(nn.Conv2d(256, 128, kernel_size=1),
                                      nn.BatchNorm2d(128),
                                      nn.ReLU(inplace=True)
                                      )
        self.D_Conv_2 = nn.Sequential(nn.Conv2d(128, 64, kernel_size=1),
                                      nn.BatchNorm2d(64),
                                      nn.ReLU(inplace=True)
                                      )
        self.D_Conv_1 = nn.Sequential(nn.Conv2d(64, 32, kernel_size=1),
                                      nn.BatchNorm2d(32),
                                      nn.ReLU(inplace=True),
                                      )

        self.Conv_4_c = nn.Sequential(nn.Conv2d(256, 128, kernel_size=1),
                                      nn.BatchNorm2d(128),
                                      nn.ReLU(inplace=True)
                                      )
        self.Conv_3_c = nn.Sequential(nn.Conv2d(128, 64, kernel_size=1),
                                      nn.BatchNorm2d(64),
                                      nn.ReLU(inplace=True)
                                      )
        self.Conv_2_c = nn.Sequential(nn.Conv2d(64, 32, kernel_size=1),
                                      nn.BatchNorm2d(32),
                                      nn.ReLU(inplace=True)
                                      )

        self.Up = nn.Upsample(scale_factor=2, mode='bilinear', align_corners=True)

        self.dps_2 = nn.Sequential(
            nn.Conv2d(64, 32, kernel_size=1),
            nn.BatchNorm2d(32),
            nn.ReLU(inplace=True),
            nn.Conv2d(32, 1, kernel_size=3, padding=1),
        )

        self.out = nn.Sequential(
            nn.Conv2d(32, 16, kernel_size=1),
            nn.BatchNorm2d(16),
            nn.ReLU(inplace=True),
            nn.Conv2d(16, 1, kernel_size=3, padding=1))

        initialize_weights(self.bfam_1, self.bfam_2, self.bfam_3, self.bfam_4,
                           self.cbm_1, self.cbm_2, self.cbm_3, self.cbm_4, self.out,
                           self.Conv_2_c, self.Conv_3_c, self.Conv_4_c,
                           self.D_Conv_2, self.D_Conv_3, self.D_Conv_4, self.D_Conv_1,
                           self.dps_2,
                           self.defm_1, self.defm_2, self.defm_3, self.defm_4
                           )

    def forward(self, inp1, inp2):
        b1 = self.reset(inp1)  # b1 的输出 shape: [1, 64, 128, 128], [1, 128, 64, 64], [1, 256, 32, 32], [1, 512, 16, 16]
        b2 = self.reset(inp2)  # b2 的输出 shape: [1, 64, 128, 128], [1, 128, 64, 64], [1, 256, 32, 32], [1, 512, 16, 16]

        # 第四层
        d4 = self.cbm_4(b1[3], b2[3])  # d4 shape: [1, 512, 16, 16]
        d4_2 = self.D_Conv_4(d4)  # d4_2 shape: [1, 256, 16, 16]
        d4_1 = self.Up(d4_2)  # d4_1 shape: [1, 256, 32, 32]

        j4 = self.bfam_4(b1[3], b2[3])  # j4 shape: [1, 256, 16, 16]

        j4_1 = self.Up(self.Conv_4_c(j4))  # j4_1 shape: [1, 128, 32, 32]
        c4 = self.defm_4(j4, d4_2)  # c4 shape: [1, 256, 32, 32]

        # 第三层
        d3 = self.cbm_3(b1[2], b2[2]) + d4_1  # d3 shape: [1, 256, 32, 32]
        d3_2 = self.D_Conv_3(d3)  # d3_2 shape: [1, 128, 32, 32]
        d3_1 = self.Up(d3_2)  # d3_1 shape: [1, 128, 64, 64]
        j3 = self.bfam_3(b1[2], b2[2], c4) + j4_1  # j3 shape: [1, 128, 32, 32]

        j3_1 = self.Up(self.Conv_3_c(j3))  # j3_1 shape: [1, 64, 64, 64]
        c3 = self.defm_3(j3, d3_2)  # c3 shape: [1, 128, 64, 64]

        # 第二层
        d2 = self.cbm_2(b1[1], b2[1]) + d3_1  # d2 shape: [1, 128, 64, 64]
        d2_2 = self.D_Conv_2(d2)  # d2_2 shape: [1, 64, 64, 64]
        d2_1 = self.Up(d2_2)  # d2_1 shape: [1, 64, 128, 128]
        j2 = self.bfam_2(b1[1], b2[1], c3) + j3_1  # j2 shape: [1, 64, 64, 64]
        j2_1 = self.Up(self.Conv_2_c(j2))  # j2_1 shape: [1, 32, 128, 128]
        c2 = self.defm_2(j2, d2_2)  # c2 shape: [1, 64, 128, 128]

        # 第一层
        d1 = self.cbm_1(b1[0], b2[0]) + d2_1  # d1 shape: [1, 64, 128, 128]
        d1_2 = self.D_Conv_1(d1)  # d1_2 shape: [1, 32, 128, 128]
        d1_1 = self.Up(d1)  # d1_1 shape: [1, 64, 256, 256]
        j1 = self.bfam_1(b1[0], b2[0], c2) + j2_1  # j1 shape: [1, 32, 128, 128]
        c1 = self.defm_1(j1, d1_2)  # c1 shape: [1, 32, 256, 256]

        # 输出层1
        out2 = self.dps_2(d1_1)  # out2 shape: [1, 1, 256, 256]

        # 最终输出层
        out = self.out(c1)  # out shape: [1, 1, 256, 256]

        return out, out

if __name__ == '__main__':
    from thop import profile, clever_format  # 根据实际模型定义路径导入

    # 创建输入张量
    inp1 = torch.rand(1, 3, 256, 256)
    inp2 = torch.rand(1, 3, 256, 256)

    # 创建模型实例
    model = B2CNet()

    # 计算输出（根据模型的前向传播）
    out, ds = model(inp1, inp2)

    # 使用 THOP 计算 FLOPS 和参数量
    flops, param = profile(model, inputs=(inp1, inp2))

    # 格式化 FLOPS 和参数量
    flops, param = clever_format([flops, param], "%.2f")

    # 显示结果
    print(f"FLOPS: {flops}")
    print(f"Parameters: {param}")
